### PR TITLE
refactor: use parseDigits() to replace strtol()

### PR DESCRIPTION
**Pull request by WalterBright**
Step toward making `parseCommandLine()` pure by encapsulating reliance on C's `errno`.

Blocked by https://github.com/dlang/dmd/pull/7289 and https://github.com/dlang/dmd/pull/7285

### CODE REVIEW

File: src/ddmd/mars.d [hunk 1]
```diff
@@ -2002,11 +2010,11 @@ private bool parseCommandLine(const ref Strings arguments, const size_t argc, re
                 {
                     if (isdigit(cast(char)p[7]))
                     {
-                        errno = 0;
-                        const level = strtol(p + 7, &p, 10);
-                        if (*p || errno || level > INT_MAX)
+                        const level = parseDigits(p + 7, int.max);
+                        if (level == uint.max)
                             goto Lerror;
-                        DebugCondition.setGlobalLevel(cast(int)level);
```
Comment (ibuclaw) [inline src/ddmd/mars.d]: Who else is using this method?

File: src/ddmd/mars.d [hunk 2]
```diff
@@ -1682,11 +1691,10 @@ private bool parseCommandLine(const ref Strings arguments, const size_t argc, re
...
                 if (p[8] == '=' && isdigit(cast(char)p[9]))
                 {
-                    errno = 0;
-                    const num = strtol(p + 9, &p, 10);
-                    if (*p || errno || num > INT_MAX)
+                    const num = parseDigits(p + 9, int.max);
+                    if (num == uint.max)
                         goto Lerror;
-                    global.errorLimit = cast(uint)num;
+                    params.errorLimit = num;
```
Comment (ibuclaw) [inline src/ddmd/mars.d]: This seems to assume that some other change has been merged.

Comment (WalterBright) [reply src/ddmd/mars.d]: That's why there's the "blocked by" note above :-)

File: src/ddmd/mars.d (hunk 1 above)
Comment (WalterBright) [reply src/ddmd/mars.d]: See https://github.com/dlang/dmd/pull/7285

File: src/ddmd/mars.d [hunk 3]
```diff
@@ -1498,6 +1493,33 @@ private bool parseCommandLine(const ref Strings arguments, const size_t argc, re
...
+     * Returns:
+     *  uint.max on error, otherwise converted integer
+     */
+    static pure uint parseDigits(const(char)*p, const uint max)
+    {
+        uint value;
+        bool overflow;
+        while (1)
+        {
+            const c = *p;
```
Comment (andralex) [inline src/ddmd/mars.d]: Save a comparison per pass, simplify flow:

```d
for (uint d; (d = uint(*p) - uint('0')) <= 9; ++p)
{
    import core.checkedint : mulu, addu;
    value = mulu(value, 10, overflow);
    value = addu(value, d, overflow);
}
return (overflow || value > max || c) ? uint.max : value;
```

**Review by andralex — APPROVED** [review]
Guess speed is not important but made a suggestion just in case.

File: src/ddmd/mars.d (hunk 3 above)
Comment (WalterBright) [reply src/ddmd/mars.d]: It's beautiful, but that `c` in the return expression is a problem. `d` has gone out of scope, so I'd have to use `*p`, meaning an extra indirection.

Comment (WalterBright) [reply src/ddmd/mars.d]: Ok, I made the change.

Comment (andralex) [reply src/ddmd/mars.d]: That's only one indirection, not one per loop. You can of course change the code like this at the cost of some minor obfuscation:

```d
uint d;
for (; (d = uint(*p) - uint('0')) <= 9; ++p)
{
    import core.checkedint : mulu, addu;
    value = mulu(value, 10, overflow);
    value = addu(value, d, overflow);
}
return (overflow || value > max || d + '0' != 0) ? uint.max : value;
```

Comment (andralex) [reply src/ddmd/mars.d]: Eh, I realized there's indeed a better version with an infinite loop:

```d
while (1)
{
    const d = uint(*p) - uint('0');
    if (d > 9)
        return uint.max;
    import core.checkedint : mulu, addu;
    value = mulu(value, 10, overflow);
    value = addu(value, d, overflow);
    ++p;
}
return (overflow || value > max) ? uint.max : value;
```

Looks like this does minimal work.

**Review by andralex — APPROVED** [review]

**Comment by andralex** [comment]
feel free to pull

**Comment by stefan-koch-sociomantic** [comment]
You use a possibly valid value as error signal ?

**Pull request merged**